dev/AppNotifications/AppNotificationBuilder/AppNotificationProgressBar.cpp

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "AppNotificationProgressBar.h"
+#include "Microsoft.Windows.AppNotifications.Builder.AppNotificationProgressBar.g.cpp"
+
+namespace winrt::Microsoft::Windows::AppNotifications::Builder::implementation
+{
+    AppNotificationProgressBar::AppNotificationProgressBar()
+        :m_titleStatus{ BindMode::NotSet},
+         m_statusStatus{ BindMode::NotSet },
+         m_valueStatus{ BindMode::NotSet },
+         m_valueStringOverrideStatus{ BindMode::NotSet }
```

Comment:
We should remove the BindMode::Enum and the member *Status postfix member variables.

For all Bind* functions, we can set the member variable to the constant string. For all Set* functions, we can set the member variable to the passed in param. 

The default ctor should set m_status/m_value to the bind values and leave the rest empty.

Reply:
What about value? how would you handle that?

Reply:
I would convert m_status from double->winrt::hstring and set the value to {progressValue}.

Reply:
But if the user sets the title property to "title", then calls TitleBind(), then retrive the Title property, he'd get {progreeTitle} but if he does the same for value, what would he get? 0.0, -1.0? we throw?

Reply:
I'm ok with throwing here.

Reply:
Whatever solution we implement, it needs to have the same behavior for users. i.e.: interacting with the title property should work in a similar way as interacting with the value property for example. Users will expect the same behavior.

Reply:
All the other properties are strings which makes them easier, but value is definitely more difficult to handle. If we don't throw, I think 0 or -1 are decent defaults to return.

Reply:
We can't use 0 because it is a valid value.

I'm not a fan of using things like -1. that seems like bad design but if the other reviewers agree that -1 is the correct solution, then I'll change it.

Reply:
> We can't use 0 because it is a valid value.
> 
> I'm not a fan of using things like -1. that seems like bad design but if the other reviewers agree that -1 is the correct solution, then I'll change it.

I'm not a fan of using valid values as 'no value'. NULL exists for a reason :)

`std::optional<int>` is the Modern C++ way to track a value that may not have a value.

Reply:
Actually a std::variant would be the right solution here, but this isn't the api we have. What we have is a double

Reply:
What should we use as the binding default double for value? @hulumane

Reply:
The toolkit sets the default to 0: https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/b21dbe71f64e86ca9bb0ba6cfb4cb4d20e93765a/Microsoft.Toolkit.Uwp.Notifications/Adaptive/AdaptiveProgressBar.cs#L39

If we do it this way, we need to track if the value was manually set to 0 by the dev since it is valid value.

Reply:
@andrewleader  as well for input.

Reply:
I'm not sure I fully understand the specific problem trying to be addressed here, but I'll try to give some general info.

Note that toast progress bars MUST have a value to be valid (I just tested a toast XML without any value attribute and got a "New notification" notification).

I believe XAML progress bars have a default value of 0, **hence why I decided to just make the default value 0**, even though technically you could argue this is a "nullable" property (but if Value doesn't have a value, you have an invalid toast). I felt just defaulting to 0 is the most intuitive and what developers would "expect", similar to how a string value typically defaults to null, and a new XAML ProgressBar "just works" without setting a value (defaults to 0).

Why would we need to track if the value was manually set to 0 by the dev? Regardless of who sets it to 0, we write it as 0, yeah? Or is this related to returning a value depending on whether BindValue() vs SetValue() was called?

Reply:
@andrewleader : As I understand the specs the value element in the xml can have a value between 0 and 1 or {progressTitle}. That all fine and dandy but we have a property that expose the value as a double. That property can represent a value between 0 and 1 but cannot represent {progressTitle}.

There are 2 options:
1- represent {progressValue} as -1. so the double property can have a value from 0 to 1 or -1.
2- Only use 0 to 1 for the property and maintain an internal flag to indicate whether the value from the property (0 to 1) or {progressValue} should be placed in the xml.